coderd/database/dbtestutil/db.go

```diff
@@ -110,3 +126,75 @@ func dbNameFromConnectionURL(t testing.TB, connectionURL string) string {
 	require.NoError(t, err)
 	return strings.TrimPrefix(u.Path, "/")
 }
+
+func DumpOnFailure(t testing.TB, connectionURL string) {
+	if !t.Failed() {
+		return
+	}
+	cwd, err := filepath.Abs(".")
+	if err != nil {
+		t.Errorf("dump on failure: cannot determine current working directory")
+		return
+	}
+	snakeCaseName := regexp.MustCompile("[^a-zA-Z0-9-_]+").ReplaceAllString(t.Name(), "_")
+	outPath := filepath.Join(cwd, snakeCaseName+".test.sql")
+	dump, err := pgDump(connectionURL)
+	if err != nil {
+		t.Errorf("dump on failure: failed to run pg_dump")
+		return
+	}
+	if err := os.WriteFile(outPath, filterDump(dump), 0o600); err != nil {
+		t.Errorf("dump on failure: failed to write: %s", err.Error())
+		return
+	}
+	t.Logf("Dumped database to %q due to failed test. I hope you find what you're looking for!", outPath)
+}
+
+// pgDump runs pg_dump against dbURL and returns the output.
+func pgDump(dbURL string) ([]byte, error) {
+	if _, err := exec.LookPath("pg_dump"); err != nil {
+		return nil, xerrors.Errorf("could not find pg_dump in path: %w", err)
+	}
+	cmdArgs := []string{
+		"pg_dump",
+		dbURL,
+		"--data-only",
+		"--column-inserts",
+		"--no-comments",
+		"--no-privileges",
+		"--no-publication",
+		"--no-security-labels",
+		"--no-subscriptions",
+		"--no-tablespaces",
+		// "--no-unlogged-table-data", // some tables are unlogged and may contain data of interest
+		"--no-owner",
+		"--exclude-table=schema_migrations",
```

Comment:
nit: should we skip more Coder tables by default?

Reply:
I'd rather err on the side of including them; schema_migrations is probably not that useful when inspecting unit test outputs though.